sql/parser/type_check.go

```diff
@@ -0,0 +1,358 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Tamir Duberstein (tamird@gmail.com)
+
+package parser
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/cockroachdb/cockroach/util"
+)
+
+// TypeCheckExpr type-checks an SQL expression. Expression checking is
+// a mostly straightforward walk over the parse tree.
+// TODO(tamird): tuple handling is currently crappy.
+func TypeCheckExpr(expr Expr) (reflect.Type, error) {
+	switch t := expr.(type) {
+	case *AndExpr:
+		return typeCheckBooleanExprs(t.Left, t.Right)
+
+	case *OrExpr:
+		return typeCheckBooleanExprs(t.Left, t.Right)
+
+	case *NotExpr:
+		return typeCheckBooleanExprs(t.Expr)
+
+	case *ParenExpr:
+		// NormalizeExpr unwraps this.
+
+	case *ComparisonExpr:
+		return typeCheckComparisonExpr(t)
+
+	case *RangeCond:
+		// NormalizeExpr transforms this into an AndExpr.
+
+	case *NullCheck:
+		_, err := TypeCheckExpr(t.Expr)
+		return boolType, err
+
+	case *ExistsExpr:
+		// The subquery within the exists should have been executed before
+		// expression evaluation and the exists nodes replaced with the result.
+
+	case BytesVal:
+		return stringType, nil
+
+	case StrVal:
+		return stringType, nil
+
+	case IntVal:
+		return intType, nil
+
+	case NumVal:
+		return floatType, nil
+
+	case BoolVal:
+		return boolType, nil
+
+	case ValArg:
+		// Placeholders should have been replaced before expression evaluation.
+
+	case *QualifiedName:
+		return nil, fmt.Errorf("qualified name \"%s\" not found", t)
+
+	case Tuple:
+		// TODO(tamird): handling of tuples leaves much to be desired. This
+		// is largely a consequence of go's type system not being
+		// expressive enough to encode the type of a heterogenous tuple
+		// (dBool, dString) as distinct from the tuple (dBool). We'll have
+		// to write our own type system here to do this properly.
+		return tupleType, nil
```

Comment:
You should at least loop over the tuple elements and verify they have the same type or the `nullType`.